### PR TITLE
Enable cancelling of running lanes

### LINKED ISSUE #11394: [Swift] Error message when lane can't be found is unclear
```
[✔] 🚀
+-------------------------------+---------+---------------+
|                      Used plugins                       |
+-------------------------------+---------+---------------+
| Plugin                        | Version | Action        |
+-------------------------------+---------+---------------+
| fastlane-plugin-emoji_fetcher | 0.2.0   | emoji_fetcher |
| fastlane-plugin-trainer       | 0.4.0   | trainer       |
+-------------------------------+---------+---------------+

[12:11:55]: FastlaneSwiftRunner project is up-to-date
[12:11:55]: $ xcodebuild -list -project ./fastlane/swift/FastlaneSwiftRunner/FastlaneSwiftRunner.xcodeproj
[12:11:56]: $ xcodebuild -showBuildSettings -scheme FastlaneRunner -project ./fastlane/swift/FastlaneSwiftRunner/FastlaneSwiftRunner.xcodeproj
[12:11:56]: $ set -o pipefail && xcodebuild -scheme FastlaneRunner -project ./fastlane/swift/FastlaneSwiftRunner/FastlaneSwiftRunner.xcodeproj -destination 'generic/platform=macOS' | tee /Users/felixkrause/Library/Logs/gym/App-FastlaneRunner.log | xcpretty
[12:11:59]: Waiting for 5 seconds for a connection from FastlaneRunner
[12:11:59]: $ ./fastlane/FastlaneRunner lane screenshots > /dev/null
[12:11:59]: Client connected
[12:11:59]: ▸ Fatal error: unable to find lane named: screenshots: file /Users/felixkrause/Developer/themoji/fastlane/swift/LaneFileProtocol.swift, line 89
[12:12:00]: Client disconnected, or a pipe broke
[12:12:00]: ▸ sh: line 1: 17755 Illegal instruction: 4  ./fastlane/FastlaneRunner lane screenshots > /dev/null
[12:12:00]: Exit status of command './fastlane/FastlaneRunner lane screenshots > /dev/null' was 132 instead of 0.
Fatal error: unable to find lane named: screenshots: file /Users/felixkrause/Developer/themoji/fastlane/swift/LaneFileProtocol.swift, line 89
sh: line 1: 17755 Illegal instruction: 4  ./fastlane/FastlaneRunner lane screenshots > /dev/null

[12:12:00]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...

bundler: failed to load command: fastlane (/Users/felixkrause/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/bin/fastlane)
FastlaneCore::Interface::FastlaneShellError: [!] Exit status of command './fastlane/FastlaneRunner lane screenshots > /dev/null' was 132 instead of 0.
Fatal error: unable to find lane named: screenshots: file /Users/felixkrause/Developer/themoji/fastlane/swift/LaneFileProtocol.swift, line 89
sh: line 1: 17755 Illegal instruction: 4  ./fastlane/FastlaneRunner lane screenshots > /dev/null

  /Users/felixkrause/Developer/fastlane/fastlane_core/lib/fastlane_core/ui/interface.rb:143:in `shell_error!'
  /Users/felixkrause/Developer/fastlane/fastlane_core/lib/fastlane_core/ui/ui.rb:14:in `method_missing'
  /Users/felixkrause/Developer/fastlane/fastlane/lib/fastlane/helper/sh_helper.rb:75:in `sh_control_output'
  /Users/felixkrause/Developer/fastlane/fastlane/lib/fastlane/helper/sh_helper.rb:12:in `sh'
  /Users/felixkrause/Developer/fastlane/fastlane/lib/fastlane/swift_lane_manager.rb:68:in `block in cruise_swift_lane_in_thread'
```

In this case, I forgot to append `Lane` at the end of the lane name my my Swift file. This is a common error, and should be handled more nicely, similar to how we handle it in the Ruby version:

```
[✔] 🚀
+-----------------------------+---------+--------------------------------+
|                              Used plugins                              |
+-----------------------------+---------+--------------------------------+
| Plugin                      | Version | Action                         |
+-----------------------------+---------+--------------------------------+
| fastlane-plugin-ruby        | 0.1.3   | rspec rubocop                  |
| fastlane-plugin-clubmate    | 0.1.0   | clubmate                       |
| fastlane-plugin-slack_train | 0.2.0   | slack_train slack_train_crash  |
|                             |         | slack_train_start              |
+-----------------------------+---------+--------------------------------+

[!] Could not find 'yolo123'. Available lanes: test, bump, generate_swift_api, release, verify_docs, update_docs, show_changelog
```

### PR DESCRIPTION
**Pull request by taquitos**
Fixes #11394

### CODE REVIEW

**Review by KrauseFx — APPROVED** [review]
I don't know enough about why this change is necessary to fix the issue with trying to run a lane that doesn't exist, could you give us some more context on what changes with this PR? I assume it's improved handling of a failed socket connection, while at the same time also adding a nice error message?

**Comment by taquitos** [comment]
@KrauseFx FastlaneRunner is launched from fastlane ruby. Once FastlaneRunner realizes it doesn't have a valid lane, it needs to talk to fastlane ruby to let it know the run is cancelled and why. That way the ruby server can shutdown and display a message. Once the ruby server acks the request, the client is safe to shutdown and exit.
This PR adds the shutdown request with reason + server ack
It's not as clean as I'd like, so I'm going to spend some time refining it.

**Comment by taquitos** [comment]
There was some cruft left over when I made a base class for all lane managers, so I deleted the duplicate code, so there is some cleanup from that you're seeing too.

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: fastlane/lib/fastlane/lane_manager.rb [hunk 1]
```diff
@@ -78,50 +77,6 @@ def self.skip_docs?
       Helper.test? || FastlaneCore::Env.truthy?("FASTLANE_SKIP_DOCS")
     end
 
-    # All the finishing up that needs to be done
```
Comment (taquitos) [inline fastlane/lib/fastlane/lane_manager.rb]: this all existed in the base class already, just cruft here.

File: fastlane/lib/fastlane/server/action_command.rb [hunk 2]
```diff
@@ -49,22 +57,5 @@ def target_class
...
       return class_name.to_s.length > 0
     end
-
-    attr_reader :command_id # always present
-    attr_reader :args # always present
-    attr_reader :method_name # always present
-    attr_reader :class_name # only present when executing a class-method
-  end
-
-  class CommandReturn
```
Comment (taquitos) [inline fastlane/lib/fastlane/server/action_command.rb]: moved to own file

File: fastlane/lib/fastlane/server/socket_server.rb [hunk 3]
```diff
@@ -146,72 +213,4 @@ def execute_command(command_json: nil)
       return "{\"payload\":{\"status\":\"failure\",\"failure_information\":#{exception_array.flatten}}}"
     end
   end
-
-  class JSONReturnValueProcessor
```
Comment (taquitos) [inline fastlane/lib/fastlane/server/socket_server.rb]: moved to own file

**Comment by taquitos** [comment]
Ok, totally updated how the server and client talk to each other. It's much more clear if you look at the files in whole, instead of as diffs. 
Now the server expects `commands` to come in, each command can be an `ActionCommand`, or a `ControlCommand`. ActionCommands tell the server to execute a fastlane action, while ControlCommands tell the server to do something like shutdown, cancel running, or clean up after finished.

When something goes wrong in the client, it now sends a cancel command which the server acks.
No more special handling for `done`, everything goes through command processing.

**Pull request merged**